python/code.qmd

```diff
@@ -0,0 +1,406 @@
+consistent:
+
+1.  An individual public-facing functionality has its own Python file.
+    Usually this means that each public function or class has its own
+    file, but if the functionality is similar enough, multiple functions or classes
+    could be grouped together in a single file.
+2.  If some internal functionality is only used by public-facing
+    functionality located in a single file, keep the internal functionality in
+    that same file.
+3.  If some internal functionality is used by multiple public-facing
+    functionalities, then you should write it in a separate file in a
+    folder called `internal/`.
+4.  If internal functionality is kept in the `internal/` folder, then
+    group the internal functions into files by their common action.
+
+Let's look at an example of how this works in practice. Let's say you
+have a package called `health_data_checks`. In this package there are
+two public functions called `check_blood_pressure()` and
+`check_heart_rate()`. Both of these functions use an internal function
+called `_is_data_object()`. The file structure might look like:
+
+```
+health-data-checks/
+└── src/
+    └── health_data_checks/
+        ├── __init__.py
+        ├── check_blood_pressure.py
+        ├── check_heart_rate.py
+        └── internals/
+            ├── __init__.py
+            └── is.py
+```
+
+Since `_is_data_object()` has an action `is`, it is put into the `is.py`
+file in the `internals/` folder. By naming the file after the action, you can add other internal functions with the same action to this file as well. This keeps the internal code
+organised around actions and also allows you to use
+`from internals import FUNCTION` when needing to use the internal
+functionality.
+
+## How to write code
+
+As in nearly all programming languages, how you write code is up to you
+as the developer. The exact details depend heavily on the problem your
+software is trying to solve, the design, the constraints, and the
+requirements.
+
+However, there are some good general practices to follow that will help
+you write more effective and maintainable Python code.
+
+### Avoiding "magic"
+
+Code that is "magic" does something unexpected or
+unpredictable. Usually this is because it has ["side
+effects"](#side-effects) or conditionally changes its action in
+fundamental ways that are not always obvious.
+
+The way you avoid magic is by writing code that is simple, predictable,
+and consistent. This means that writing code in a more [functional
+programming](https://en.wikipedia.org/wiki/Functional_programming) style
+tends to produce more maintainable and more predictable code. Strategies
+for writing more functional code include:
+
+-   Using clear inputs and one type of output for functions and methods.
+    There is more explanation on this in the [type hints](#type-hints)
+    section below.
+-   Avoiding [side effects](#side-effects) by only outputting the result of the function
```

Comment:
Should this have a reference to the section on side effects in the same way as type hints above?